### PR TITLE
support year field

### PR DESCRIPTION
**Pull request by JabberWocky-22**
support year at the seventh field.

### CODE REVIEW

File: src/croniter/croniter.py [hunk 1]
```diff
@@ -790,11 +823,12 @@ def _expand(cls, expr_format, hash_id=None, second_at_beginning=False, from_time
...
                     if t in cls.LOWMAP[i] and not (
-                        # do not support 0 as a month either for classical 5 fields cron
-                        # or 6fields second repeat form
+                        # do not support 0 as a month either for classical 5 fields cron,
+                        # 6fields second repeat form or 7 fields year form
                         # but still let conversion happen if day field is shifted
                         (i in [2, 3] and len(expressions) == 5) or
-                        (i in [3, 4] and len(expressions) == 6)
+                        (i in [3, 4] and len(expressions) == 6) or
+                        (i in [2, 3, 4] and len(expressions) == 7)
```
Comment (JabberWocky-22) [inline src/croniter/croniter.py:831]: I have no idea why should check second or forth field depending on length of expressions.
Please correct this.

**Comment by kiorky** [comment]
i'm willing to accept this, but i have to review it carefully.

Please make sure to add extensive tests, cron is devil in case of matrix cases ;-)

File: src/croniter/croniter.py [hunk 2]
```diff
@@ -378,19 +383,37 @@ def _calc(self, now, expanded, nth_weekday_of_month, is_prev):
...
             nearest_diff_method = self._get_prev_nearest_diff
             sign = -1
-            offset = (len(expanded) == 6 or now % 60 > 0) and 1 or 60
+            offset = (len(expanded) >= 6 or now % 60 > 0) and 1 or 60
         else:
             now = math.floor(now)
             nearest_diff_method = self._get_next_nearest_diff
             sign = 1
-            offset = (len(expanded) == 6) and 1 or 60
+            offset = (len(expanded) >= 6) and 1 or 60
```
Comment (kiorky) [inline src/croniter/croniter.py]: `> UNIX_CRON_LEN`

File: src/croniter/croniter.py [hunk 3]
```diff
@@ -539,7 +562,7 @@ def proc_minute(d):
             return False, d
 
         def proc_second(d):
-            if len(expanded) == 6:
+            if len(expanded) >= 6:
```
Comment (kiorky) [inline src/croniter/croniter.py]: `> UNIX_CRON_LEN`

File: src/croniter/croniter.py [hunk 4]
```diff
@@ -668,9 +701,9 @@ def _expand(cls, expr_format, hash_id=None, second_at_beginning=False, from_time
         if len(expressions) not in VALID_LEN_EXPRESSION:
             raise CroniterBadCronError(cls.bad_length)
 
-        if len(expressions) == 6 and second_at_beginning:
+        if len(expressions) >= 6 and second_at_beginning:
```
Comment (kiorky) [inline src/croniter/croniter.py]: `> UNIX_CRON_LEN`

File: src/croniter/tests/test_croniter.py [hunk 5]
```diff
@@ -894,7 +894,7 @@ def test_error_bad_cron(self):
         self.assertRaises(CroniterBadCronError, croniter.expand,
                           '* * * *')
         self.assertRaises(CroniterBadCronError, croniter.expand,
-                          '* * * * * * *')
+                          '* * * * * * * *')
```
Comment (kiorky) [inline src/croniter/tests/test_croniter.py]: rebase, i just changed this test to be dynamic, you wont have to do anything there.

File: src/croniter/croniter.py [hunk 6]
```diff
@@ -378,19 +383,37 @@ def _calc(self, now, expanded, nth_weekday_of_month, is_prev):
...
         month, year = dst.month, dst.year
         current_year = now.year
         DAYS = self.DAYS
 
+        def proc_year(d):
+            if len(expanded) == 7:
+                try:
+                    expanded[6].index("*")
+                except ValueError:
+                    diff_year = nearest_diff_method(d.year, expanded[6], None)
```
Comment (kiorky) [inline src/croniter/croniter.py]: expanded[YEAR_FIELD]

File: src/croniter/croniter.py [hunk 7]
```diff
@@ -378,19 +383,37 @@ def _calc(self, now, expanded, nth_weekday_of_month, is_prev):
...
         dst = now = self._timestamp_to_datetime(now + sign * offset)
 
         month, year = dst.month, dst.year
         current_year = now.year
         DAYS = self.DAYS
 
+        def proc_year(d):
+            if len(expanded) == 7:
+                try:
+                    expanded[6].index("*")
```
Comment (kiorky) [inline src/croniter/croniter.py]: expanded[YEAR_FIELD]

File: src/croniter/croniter.py [hunk 8]
```diff
@@ -378,19 +383,37 @@ def _calc(self, now, expanded, nth_weekday_of_month, is_prev):
...
+            offset = (len(expanded) >= 6) and 1 or 60
 
         dst = now = self._timestamp_to_datetime(now + sign * offset)
 
         month, year = dst.month, dst.year
         current_year = now.year
         DAYS = self.DAYS
 
+        def proc_year(d):
+            if len(expanded) == 7:
```
Comment (kiorky) [inline src/croniter/croniter.py]: YEAR_CRON_LEN

File: src/croniter/croniter.py [hunk 9]
```diff
@@ -668,9 +701,9 @@ def _expand(cls, expr_format, hash_id=None, second_at_beginning=False, from_time
         if len(expressions) not in VALID_LEN_EXPRESSION:
             raise CroniterBadCronError(cls.bad_length)
 
-        if len(expressions) == 6 and second_at_beginning:
+        if len(expressions) >= 6 and second_at_beginning:
             # move second to last to process by same logical
-            expressions.append(expressions.pop(0))
+            expressions.insert(5, expressions.pop(0))
```
Comment (kiorky) [inline src/croniter/croniter.py]: insert(SECOND_FIELD,

File: src/croniter/tests/test_croniter_hash.py [hunk 10]
```diff
@@ -420,7 +461,7 @@ def test_expand_month_days_range_3_months(self):
         assert min(months) == self.MIN_VALUE
         assert max(months) == self.MAX_VALUE
 
-    def test_expand_month_days_range_5_months(self):
+    def test_expand_months_range_5_months(self):
```
Comment (kiorky) [inline src/croniter/tests/test_croniter_hash.py]: please do not change.

File: src/croniter/tests/test_croniter_hash.py [hunk 11]
```diff
@@ -408,7 +449,7 @@ def test_expand_month_days_range_2_months(self):
         assert min(months) == self.MIN_VALUE
         assert max(months) == self.MAX_VALUE
 
-    def test_expand_month_days_range_3_months(self):
+    def test_expand_months_range_3_months(self):
```
Comment (kiorky) [inline src/croniter/tests/test_croniter_hash.py]: please do not change.

File: src/croniter/tests/test_croniter_hash.py [hunk 12]
```diff
@@ -396,7 +437,7 @@ def test_expand_month_days(self):
         assert min(month_days) == self.MIN_VALUE
         assert max(month_days) == self.MAX_VALUE
 
-    def test_expand_month_days_range_2_months(self):
+    def test_expand_months_range_2_months(self):
```
Comment (kiorky) [inline src/croniter/tests/test_croniter_hash.py]: please do not change.

File: src/croniter/tests/test_croniter_hash.py [hunk 13]
```diff
@@ -380,13 +411,23 @@ def test_expand_month_days_range_12_days(self):
...
+        assert max(month_days) == self.MAX_VALUE
+
 
 class CroniterHashExpanderExpandMonthTest(CroniterHashExpanderBase):
     MIN_VALUE = 1
     MAX_VALUE = 12
     TOTAL = 12
 
-    def test_expand_month_days(self):
+    def test_expand_months(self):
```
Comment (kiorky) [inline src/croniter/tests/test_croniter_hash.py]: please do not change.

File: src/croniter/tests/test_croniter_hash.py [hunk 14]
```diff
@@ -146,7 +159,7 @@ def test_hash_word_midnight(self):
 
         @midnight is actually up to 3 hours after midnight, not exactly midnight
         """
-        self._test_iter('@midnight', datetime(2020, 1, 1, 1, 10, 32), timedelta(days=1))
+        self._test_iter('@midnight', datetime(2020, 1, 1, 2, 10, 32), timedelta(days=1))
```
Comment (kiorky) [inline src/croniter/tests/test_croniter_hash.py]: should not change, bug introduced.

**Comment by JabberWocky-22** [comment]
I have rebased to master, fixed the test and added more. Please take some time to review :) @kiorky 

Besides:
I also fixed a bug which https://github.com/kiorky/croniter/pull/42 doesn't cover(range)
Ensure `match_range` return `False` when not previous value avaliable
Sorry for mix them up.

**Comment by kiorky** [comment]
> I have rebased to master, fixed the test and added more. Please take some time to review :) @kiorky
> 
> Besides: I also fixed a bug which #42 doesn't cover(range) Ensure `match_range` return `False` when not previous value avaliable Sorry for mix them up.

yes i saw, this worries me, can you make a separate PR for this, i ll merge it way faster that this one that should be scopped as it is already invasive.

**Comment by kiorky** [comment]
but how did you rebase, as there are 3 commits, can you just make a PR with the match_range problem, then rebase (as `git rebase -i`) on top of it ?

**Comment by JabberWocky-22** [comment]
I have made 2 separate bug fix PR.

I just rebase this PR to master and didn't squash the commits. Some projects dislike squash and force-push.
I will squash it after it's ok.

File: src/croniter/croniter.py [hunk 15]
```diff
@@ -816,11 +849,12 @@ def _expand(cls, expr_format, hash_id=None, second_at_beginning=False, from_time
...
-                        # do not support 0 as a month either for classical 5 fields cron
-                        # or 6fields second repeat form
+                        # do not support 0 as a month either for classical 5 fields cron,
+                        # 6fields second repeat form or 7 fields year form
                         # but still let conversion happen if day field is shifted
-                        (i in [2, 3] and len(expressions) == 5) or
-                        (i in [3, 4] and len(expressions) == 6)
+                        (i in [DAY_FIELD, MONTH_FIELD] and len(expressions) == UNIX_CRON_LEN) or
+                        (i in [MONTH_FIELD, DOW_FIELD] and len(expressions) == SECOND_CRON_LEN) or
+                        (i in [DAY_FIELD, MONTH_FIELD, DOW_FIELD] and len(expressions) == YEAR_CRON_LEN)
```
Comment (kiorky) [inline src/croniter/croniter.py:877]: Well those tests are maybe broken (not from your, initially), i have to rethink about this...

File: src/croniter/croniter.py [hunk 16]
```diff
@@ -929,12 +963,15 @@ def match_range(cls, cron_expression, from_datetime, to_datetime,
                     day_or=True, second_at_beginning=False):
         cron = cls(cron_expression, to_datetime, ret_type=datetime.datetime,
                    day_or=day_or, second_at_beginning=second_at_beginning)
-        td, ms1 = cron.get_current(datetime.datetime), relativedelta(microseconds=1)
-        if not td.microsecond:
-            td = td + ms1
-        cron.set_current(td, force=True)
-        tdp, tdt = cron.get_current(), cron.get_prev()
-        precision_in_seconds = 1 if len(cron.expanded) == 6 else 60
+        tdp = cron.get_current(datetime.datetime)
```
Comment (kiorky) [inline src/croniter/croniter.py]: block from another PR already merge/reviewed to be rebased

**Comment by kiorky** [comment]
Damn, i m so sorry, seems the review i already  did yesterday wasnt published.
So some (yesterday) comments may be out of date.

What i propose to you is to make the other PR #82 merged, then you rebase this PR, then i ll ensure the review is up to date ...

**Comment by kiorky** [comment]
So please now rebase this one to see the final diff that should have a lift :)

**Comment by kiorky** [comment]
(you should obtain something similar to https://github.com/kiorky/croniter/compare/master...y )

File: src/croniter/croniter.py [hunk 17]
```diff
@@ -617,13 +647,14 @@ def _get_prev_nearest(self, x, to_check):
...
-            if d == "l":
+            if d == "l" and range_val is not None:
                 # if 'l' then it is the last day of month
                 # => its value of range_val
                 d = range_val
             if d >= x:
                 return d - x
-        return to_check[0] - x + range_val
+        if range_val is not None:
+            return to_check[0] - x + range_val
```
Comment (kiorky) [inline src/croniter/croniter.py]: what are you trying to do/solve/address with these changes toward `range_val` ?
would something not to put in a separate PR ?

Comment (JabberWocky-22) [reply src/croniter/croniter.py]: The year field doesn't have a range.
If current second is 40, and expected second is 10-20, then I can move to next minute.
But if no years match the giving value, then there is no available values anymore, since there is no loop for year field.
So I simply set `range_val` to None to indicate that shouldn't look for next loop.

Comment (JabberWocky-22) [reply src/croniter/croniter.py]: Do you need a further explanation? @kiorky

Comment (kiorky) [reply src/croniter/croniter.py]: Nope, just didn't had time in the last days, can you please add comments to document year processing tricks ? and i think we will be nearly good.

File: src/croniter/croniter.py [hunk 18]
```diff
@@ -633,6 +664,8 @@ def _get_prev_nearest_diff(self, x, to_check, range_val):
                 return d - x
         if 'l' in candidates:
             return -x
+        if range_val is None:
```
Comment (kiorky) [inline src/croniter/croniter.py:687]: also add documentation comment

File: src/croniter/croniter.py [hunk 19]
```diff
@@ -617,13 +647,14 @@ def _get_prev_nearest(self, x, to_check):
 
     def _get_next_nearest_diff(self, x, to_check, range_val):
         for i, d in enumerate(to_check):
-            if d == "l":
+            if d == "l" and range_val is not None:
```
Comment (kiorky) [inline src/croniter/croniter.py:659]: Add comment for use.

File: src/croniter/croniter.py [hunk 20]
```diff
@@ -587,12 +611,18 @@ def proc_second(d):
 
         while abs(year - current_year) <= self._max_years_between_matches:
             next = False
+            stop = False
             for proc in procs:
                 (changed, dst) = proc(dst)
+                if changed is None:
```
Comment (kiorky) [inline src/croniter/croniter.py:620]: document that is an early stop in case of year nonmatch

File: src/croniter/croniter.py (hunk 17 above)
Comment (JabberWocky-22) [reply src/croniter/croniter.py]: Thanks for your review! XD

File: src/croniter/croniter.py (hunk 18 above)
Comment (JabberWocky-22) [reply src/croniter/croniter.py:687]: I added documents to both functions. You can edit them however you want.
I'm not good at writing documents.

Comment (kiorky) [reply src/croniter/croniter.py:687]: i mean comments, document what you did, and why

File: src/croniter/croniter.py [hunk 21]
```diff
@@ -694,9 +745,9 @@ def _expand(cls, expr_format, hash_id=None, second_at_beginning=False, from_time
         if len(expressions) not in VALID_LEN_EXPRESSION:
             raise CroniterBadCronError(cls.bad_length)
 
-        if len(expressions) == 6 and second_at_beginning:
-            # move second to last to process by same logical
-            expressions.append(expressions.pop(0))
+        if len(expressions) > UNIX_CRON_LEN and second_at_beginning:
+            # move second to it's own field to process by same logical
```
Comment (kiorky) [inline src/croniter/croniter.py]: replace s/own/own (6th)/

File: src/croniter/croniter.py (hunk 18 above)
Comment (kiorky) [reply src/croniter/croniter.py:687]: I propose you the bellow modification than you can take mostly as-is (untested, verify tests :)):

- I moved a part of the docstring to where it is effectively used.

Thus, this enhances code readability, while doing the same afterall.

```python


    def _get_prev_nearest_diff(self, x, to_check, range_val):
        """
        `range_val` is the range of a field.
        If no available time, we can move to previous loop(like previous month).
        Range_val can also be set to `None` to indicate that there is no loop.
        ( Currently should only used for `year` field )
        """
        candidates = to_check[:]
        candidates.reverse()
        for d in candidates:
            if d != 'l' and d <= x:
                return d - x
        if 'l' in candidates:
            return -x

        # If no available time, we can move to previous loop (like previous month).
        # Range_val can also be set to `None` to indicate that there is no loop.
        # Currently should only used for `year` field
        # When range_val is None and x not exists in to_check,
        # `None` will be returned to suggest no more available time
        if range_val is None:
```

File: src/croniter/croniter.py (hunk 19 above)
Comment (kiorky) [reply src/croniter/croniter.py:659]: I can do it myself but I prefer you to do it for you to keep all this big PR paternity :)

I propose you the bellow modification than you can take mostly as-is (untested, verify tests :)):
- I moved a part of the docstring to where it is effectively used.
- I modified a bit the comment.
- I added a final explicit `return None`

Thus, this enhances code readability, while doing the same afterall.

```python
    def _get_next_nearest_diff(self, x, to_check, range_val):
        """
        `range_val` is the range of a field.
        If no available time, we can move to next loop(like next month).
        `range_val` can also be set to `None` to indicate that there is no loop.
        Currently, it is only used for `year` field        
        """
        for i, d in enumerate(to_check):
            if d == "l" and range_val is not None:
                # if 'l' then it is the last day of month
                # => its value of range_val
                d = range_val
            if d >= x:
                return d - x
        # When range_val is None and x not exists in to_check,
        # `None` will be returned to suggest no more available time
        if range_val is None:
             return None
        return to_check[0] - x + range_val
```

File: src/croniter/croniter.py (hunk 20 above)
Comment (kiorky) [reply src/croniter/croniter.py:620]: Add a comment like 
```python
# `None` can be set mostly for year processing 
# so please see proc_year / _get_prev_nearest_diff / _get_next_nearest_diff
```

**Comment by kiorky** [comment]
There is a bug with the comment posts, give me a minute to sort it out ...

**Comment by kiorky** [comment]
> There is a bug with the comment posts, give me a minute to sort it out ...

you can proceed, sorry for the noise :)

File: src/croniter/croniter.py (hunk 18 above)
Comment (JabberWocky-22) [reply src/croniter/croniter.py:687]: thanks for your detailed suggestions

**Comment by kiorky** [comment]
So, so far so good, merging, and a lot of thanks for applying those review modifications, along to the 2 other PRs !

**Comment by kiorky** [comment]
released as 3.0.0  https://pypi.org/project/croniter/3.0.0/
